Step 1 of cspyce reorganization

fyellin opened the pull request:
cspyce0.py moved to cspyce/ directory;  cspyce0 subdirectory renamed swig
make-vectorize.py looks only at cspyce0.i and doesn't need any helper files.
cspyce2.py is now a normal Python file and not auto generated.

cspyce0_template.i now works for both Py2 and Py3.  It generates a C file and a .py file that can work with either of them (and hence no need for -py3).

First attempts at making the code pip installable

rfrenchseti commented on cspyce/__init__.py:
```diff
@@ -13,18 +13,30 @@
+except ImportError as err:
+    if os.getenv("CSPICE_DEVELOPMENT"):
+        print("Package cspyce2 not found.  Error ignored.")
+        pass
```
This pass is not needed

fyellin replied:
Removed

rfrenchseti commented on cspyce/__init__.py:
```diff
@@ -13,18 +13,30 @@
+        print("Package cspyce2 not found.  Error ignored.")
+        pass
+    else:
+        print("Set environment variable 'CSPICE_DEVELOPMENT' to 'True' to ignore this error")
```
I don't think you can set the environment variable to "True". You have to set it to "1", right?

fyellin replied:
Anything that isn't the empty string will do.  I fixed this for you.

rfrenchseti commented on README-developers.md:
```diff
@@ -0,0 +1,113 @@
+## SETTING UP
+
+The root of the github tree is called `pds-cspyce`.
+When we use `pds-cspcyce` in this document, we are referring to that directory,
```
pds-cspyce is spelled wrong

rfrenchseti commented on README-developers.md:
```diff
@@ -0,0 +1,113 @@
+
+The github source code intentionally does not include the source code to
+cspice itself.
+You should download the cspice sources from XXXX.
```
You download CSPICE from: https://naif.jpl.nasa.gov/naif/toolkit.html

You might want to tell them which version to download, if it matters

fyellin replied:
Fixed.

rfrenchseti commented on README-developers.md:
```diff
@@ -0,0 +1,113 @@
+`pds-cspyce/cspice/src/cspice` contains several thousand `.c` files.
+
+## BUILDING CSPYCE
+Whenever I use the program `python` as a command, I mean your specific Python
```
Use "we" instead of "I".

fyellin replied:
Done.

rfrenchseti commented on cspyce/__init__.py:
```diff
@@ -180,8 +213,8 @@ def get_all_funcs(source=None, cspyce_dict=None):
 def get_all_versions(func, source=None):
-    """Return a dictionary of all cspyce functions associated with this one,
-    keyed by their names.
+    """A dictionary of all cspyce functions associated with this one, keyed by
```
I'm not sure why you changed all these docstrings. The new ones don't agree with PEP257: 

```
The docstring is a phrase ending in a period. It prescribes the function or method’s effect as a command (“Do this”, “Return that”), not as a description; e.g. don’t write “Returns the pathname …”.
```

Since this will be a public library, we should follow PEP257, including the requirement for the first line to be a single phrase that doesn't wrap to another line.

fyellin replied:
I didn't touch these.  According to my git history, these were changed before I ever saw it.

rfrenchseti commented on cspyce/cspyce1.py:
```diff
@@ -106,9 +108,12 @@
 import textwrap
 
-import cspyce.cspyce0 as cspyce0
+# import cspyce.cspyce0 as cspyce0
```
Do we need these commented-out lines?

fyellin replied:
I made no changes to cspyce1.py.  
But I deleted this

rfrenchseti commented on setup.py:
```diff
@@ -0,0 +1,111 @@
+# be able to simply install swig.
+#
+#
+# If you are doing a build from sources, you must also have have swig and a copy of
```
"have have"

fyellin replied:
I changed this whole thing, telling them to check out the README.developer isntead

rfrenchseti commented on setup.py:
```diff
@@ -0,0 +1,111 @@
+# the cspice toolkit.
+#
+# [I need appropriate instructions here on how to get cspice.a and install it. so that
+#    <this top level directory>/cspice/lib/libcspice.a
```
Actually the cspice distribution from NAIF calls the file `cspice/lib/cspice.a`, at least in the Linux version. Perhaps that was part of the problem I was having? The URL I gave earlier for downloading the Toolkit includes a pre-compiled version for each supported platform. You can just download each version, untar it, and see what's in it.

fyellin replied:
This gets back to my original issue.  I have no way of knowing where the user is going to put their libcspice.a. And setup.py doesn't give me an easy way of looking for it on the user's machine.  That's why I decided the simplest solution was just to create it myself.

rfrenchseti commented on setup.py:
```diff
@@ -0,0 +1,111 @@
+# [I need appropriate instructions here on how to get cspice.a and install it. so that
+#    <this top level directory>/cspice/lib/libcspice.a
+#    <this top level directory>/cspice/include/
+# contains the library and and the include files]
```
"and and"

fyellin replied:
Fixed.

rfrenchseti commented on setup.py:
```diff
@@ -0,0 +1,111 @@
+    'cspyce.cspice',
+    sources=glob("cspice/src/cspice/*.c"),
+    include_dirs=['cspice/include'],
+    extra_compile_args=['-Wno-incompatible-pointer-types', '-Wno-parentheses', 
```
It's going to be really interesting to try this on Windows, where these compiler flags don't work.

fyellin replied:
Yeah.  The cspice source code is sort of scary with all the the problems that the C compiler notices.  I wonder if the C compiler is uncovering some actual bugs.

fyellin commented:
On Fri, Mar 11, 2022 at 2:50 PM Robert French ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In cspyce/__init__.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825162963>:
>
> >  # We allow the import of cspyce2 to fail because, during development, there are
>
>  # times when cspyce2.py might be invalid. If that happens, we still want to be
>
>  # able to work inside the cspyce directory tree. Without allowing this
>
>  # exception, it could become impossible to work on and test cspyce functions
>
>  # inside this directory tree. This should never occur during a normal run.
>
>
>
>  try:
>
> -    from cspyce2 import *
>
> -except ImportError:
>
> -    pass
>
> +    from .cspyce2 import *
>
> +except ImportError as err:
>
> +    if os.getenv("CSPICE_DEVELOPMENT"):
>
> +        print("Package cspyce2 not found.  Error ignored.")
>
> +        pass
>
>
> This pass is not needed
>

Done


> ------------------------------
>
> In cspyce/__init__.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825163111>:
>
> >  # We allow the import of cspyce2 to fail because, during development, there are
>
>  # times when cspyce2.py might be invalid. If that happens, we still want to be
>
>  # able to work inside the cspyce directory tree. Without allowing this
>
>  # exception, it could become impossible to work on and test cspyce functions
>
>  # inside this directory tree. This should never occur during a normal run.
>
>
>
>  try:
>
> -    from cspyce2 import *
>
> -except ImportError:
>
> -    pass
>
> +    from .cspyce2 import *
>
> +except ImportError as err:
>
> +    if os.getenv("CSPICE_DEVELOPMENT"):
>
> +        print("Package cspyce2 not found.  Error ignored.")
>
> +        pass
>
> +    else:
>
> +        print("Set environment variable 'CSPICE_DEVELOPMENT' to 'True' to ignore this error")
>
>
> I don't think you can set the environment variable to "True". You have to
> set it to "1", right?
>

You can set an environment variable to anything you want.  Any value will
actually work.  But I changed it to "1"

> ------------------------------
>
> In README-developers.md
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825163350>:
>
> > @@ -0,0 +1,113 @@
>
> +# `cspyce` DEVELOPERS' GUIDE
>
> +March 2022
>
> +
>
> +Frank Yellin, PDS Ring-Moon Systems Node, SETI Institute volunteer.
>
> +
>
> +This guide is intended for those who have downloaded the
>
> +[github source code](https://github.com/SETI/pds-cspyce/) and are modifying
>
> +or extending the `cspyce` toolkit.
>
> +You do not need to read this if you are installing an existing `PyPI`
>
> +distribition of the sources.
>
> +
>
> +## SETTING UP
>
> +
>
> +The root of the github tree is called `pds-cspyce`.
>
> +When we use `pds-cspcyce` in this document, we are referring to that directory,
>
>
> pds-cspyce is spelled wrong
>

Fixed

> ------------------------------
>
> In README-developers.md
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825163844>:
>
> > +
>
> +This guide is intended for those who have downloaded the
>
> +[github source code](https://github.com/SETI/pds-cspyce/) and are modifying
>
> +or extending the `cspyce` toolkit.
>
> +You do not need to read this if you are installing an existing `PyPI`
>
> +distribition of the sources.
>
> +
>
> +## SETTING UP
>
> +
>
> +The root of the github tree is called `pds-cspyce`.
>
> +When we use `pds-cspcyce` in this document, we are referring to that directory,
>
> +wherever it happens to be on your computer.
>
> +
>
> +The github source code intentionally does not include the source code to
>
> +cspice itself.
>
> +You should download the cspice sources from XXXX.
>
>
> You download CSPICE from: https://naif.jpl.nasa.gov/naif/toolkit.html
>
> You might want to tell them which version to download, if it matters
>

I'm absolutely hoping it doesn't matter.  I just downloaded the latest
version.


> ------------------------------
>
> In README-developers.md
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825164054>:
>
> > +The root of the github tree is called `pds-cspyce`.
>
> +When we use `pds-cspcyce` in this document, we are referring to that directory,
>
> +wherever it happens to be on your computer.
>
> +
>
> +The github source code intentionally does not include the source code to
>
> +cspice itself.
>
> +You should download the cspice sources from XXXX.
>
> +You should make a soft link from your cspice directory to `pds-cspyce/cspice`.
>
> +```shell
>
> +ln -s <path-to-your-cspice-directory> <path-to-your-pds-cspyce>/cspice
>
> +```
>
> +Confirm that `pds-cspyce/cspice/include` has a bunch of `.h` files and that
>
> +`pds-cspyce/cspice/src/cspice` contains several thousand `.c` files.
>
> +
>
> +## BUILDING CSPYCE
>
> +Whenever I use the program `python` as a command, I mean your specific Python
>
>
> Use "we" instead of "I".
>

I've rewritten this entire section.  I'll change all my I's to we's.

> ------------------------------
>
> In cspyce/__init__.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825165021>:
>
> > @@ -180,8 +213,8 @@ def get_all_funcs(source=None, cspyce_dict=None):
>
>      return cspyce_dict
>
>
>
>  def get_all_versions(func, source=None):
>
> -    """Return a dictionary of all cspyce functions associated with this one,
>
> -    keyed by their names.
>
> +    """A dictionary of all cspyce functions associated with this one, keyed by
>
>
> I'm not sure why you changed all these docstrings. The new ones don't
> agree with PEP257:
>

Any changes in this file are Mark's.  My original sources were his, not
what came down from github.

> The docstring is a phrase ending in a period. It prescribes the function or method’s effect as a command (“Do this”, “Return that”), not as a description; e.g. don’t write “Returns the pathname …”.
>
>
> Since this will be a public library, we should follow PEP257, including
> the requirement for the first line to be a single phrase that doesn't wrap
> to another line.
> ------------------------------
>
> In cspyce/cspyce1.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825173702>:
>
> > @@ -106,9 +108,12 @@
>
>  import numpy as np
>
>  import textwrap
>
>
>
Again.  I didn't touch this file.  This is all Mark's doing.   WIll fix

> -import cspyce.cspyce0 as cspyce0
>
> +# import cspyce.cspyce0 as cspyce0
>
>
> Do we need these commented-out lines?
> ------------------------------
>
> In setup.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825174276>:
>
> > @@ -0,0 +1,111 @@
>
> +#!/usr/bin/env python
>
> +
>
> +# If a binary distribution exists for your OS and Python, congratulations.  You should
>
> +# be able to simply install swig.
>
> +#
>
> +#
>
> +# If you are doing a build from sources, you must also have have swig and a copy of
>
>
> "have have"
>

Ugg.  I think I successfully avoided reading those paragraphs.

> ------------------------------
>
> In setup.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825174796>:
>
> > @@ -0,0 +1,111 @@
>
> +#!/usr/bin/env python
>
> +
>
> +# If a binary distribution exists for your OS and Python, congratulations.  You should
>
> +# be able to simply install swig.
>
> +#
>
> +#
>
> +# If you are doing a build from sources, you must also have have swig and a copy of
>
> +# the cspice toolkit.
>
> +#
>
> +# [I need appropriate instructions here on how to get cspice.a and install it. so that
>
> +#    <this top level directory>/cspice/lib/libcspice.a
>
>
> Actually the cspice distribution from NAIF calls the file
> cspice/lib/cspice.a, at least in the Linux version. Perhaps that was part
> of the problem I was having? The URL I gave earlier for downloading the
> Toolkit includes a pre-compiled version for each supported platform. You
> can just download each version, untar it, and see what's in it.
> ------------------------------
>
> In setup.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825174849>:
>
> > @@ -0,0 +1,111 @@
>
> +#!/usr/bin/env python
>
> +
>
> +# If a binary distribution exists for your OS and Python, congratulations.  You should
>
> +# be able to simply install swig.
>
> +#
>
> +#
>
> +# If you are doing a build from sources, you must also have have swig and a copy of
>
> +# the cspice toolkit.
>
> +#
>
> +# [I need appropriate instructions here on how to get cspice.a and install it. so that
>
> +#    <this top level directory>/cspice/lib/libcspice.a
>
> +#    <this top level directory>/cspice/include/
>
> +# contains the library and and the include files]
>
>
> "and and"
> ------------------------------
>
> In setup.py
> <https://github.com/SETI/pds-cspyce/pull/3#discussion_r825175213>:
>
> > +            from swig.make_vectorize import create_vectorize_header_file
>
> +            create_vectorize_header_file("swig/vectorize.i")
>
> +            print("Rerunning swig")
>
> +            command = "swig -python -outdir cspyce/. -o swig/cspyce0_wrap.c swig/cspyce0.i".split(' ')
>
> +            subprocess.check_call(command)
>
> +            command = "swig -python -outdir cspyce/. -o swig/typemap_samples_wrap.c swig/typemap_samples.i".split(' ')
>
> +            subprocess.check_call(command)
>
> +        else:
>
> +            command = "echo Cannot rebuild files in Python2".split(' ')
>
> +            subprocess.check_call(command)
>
> +
>
> +cspice_module = Extension(
>
> +    'cspyce.cspice',
>
> +    sources=glob("cspice/src/cspice/*.c"),
>
> +    include_dirs=['cspice/include'],
>
> +    extra_compile_args=['-Wno-incompatible-pointer-types', '-Wno-parentheses',
>
>
> It's going to be really interesting to try this on Windows, where these
> compiler flags don't work.
>
> —
>
> There's not a snowball's chance in hell of this working on Windows.  The
basic setup should work, but I have no idea which compiler flags to use.
The cspice source code has a lot of warnings, some of which really have me
worried.  Does the cspice team have any interest in fixing them?

The most worrying is that there is some code of the form  "a - b << i"
which should never be written without parentheses.

> Message ID: ***@***.***>
>
